kotlin-codegen/integration-test/src/test/kotlin/com/squareup/moshi/GeneratedAdaptersTest.kt

```diff
@@ -378,7 +378,7 @@ class GeneratedAdaptersTest {
       jsonAdapter.fromJson("{\"a\":null}")
       fail()
     } catch (expected: JsonDataException) {
-      assertThat(expected).hasMessage("Required property 'a' missing at \$")
+      assertThat(expected).hasMessage("Unexpected null at \$.a")
```

Comment:
slight hitch with adapters that may return null (instead of a null literal in the JSON).

```kotlin
@Test fun nonNullConstructorParameterCalledWithNullFailsWithJsonDataException() {
  val moshi = Moshi.Builder().add(object {
    @FromJson fun fromJson(string: String): String? = null
  }).build()
  val jsonAdapter = moshi.adapter(HasNonNullConstructorParameter::class.java)

  try {
    jsonAdapter.fromJson("{\"a\":\"hello\"}")
    fail()
  } catch (expected: JsonDataException) {
    assertThat(expected).hasMessage("Unexpected null at \$.a")
  }
}
```

the above test case still fails with `Required property 'a' missing at $` which is not the correct error.

in the reflective adapter, in fails with `Non-null value 'a' was null at $` which makes more sense IMO.

Reply:
Agreed that it’s less legible. Making it more legible is a pain; I need to know the _property_ name instead of just the JSON path. I’m taking a shortcut and not doing that.

Reply:
hm, I'll take a look when I get a chance.

Reply:
for posterity #511.